pkg/logql/evaluator.go

```diff
@@ -325,3 +327,180 @@ func (ev *defaultEvaluator) rangeAggEvaluator(ctx context.Context, expr *rangeAg
 
 	}, vecIter.Close)
 }
+
+func (ev *defaultEvaluator) binOpEvaluator(
+	ctx context.Context,
+	expr *binOpExpr,
+	q Params,
+) (StepEvaluator, error) {
+	lhs, err := ev.Evaluator(ctx, expr.SampleExpr, q)
+	if err != nil {
+		return nil, err
+	}
+	rhs, err := ev.Evaluator(ctx, expr.RHS, q)
+	if err != nil {
+		return nil, err
+	}
+
+	return newStepEvaluator(func() (bool, int64, promql.Vector) {
+		pairs := map[uint64][2]*promql.Sample{}
+		var ts int64
+
+		// populate pairs
+		for i, eval := range []StepEvaluator{lhs, rhs} {
+			next, timestamp, vec := eval.Next()
+			ts = timestamp
+
+			// These should _always_ happen at the same step on each evaluator.
+			if !next {
+				return next, ts, nil
+			}
+
+			for _, sample := range vec {
+				// TODO(owen-d): this seems wildly inefficient: we're calculating
```

Comment:
Good point.

Reply:
opened https://github.com/grafana/loki/issues/1672 for this